services/external/google/user-client.ts

```diff
@@ -0,0 +1,134 @@
+import { logger } from "@/lib/logger";
+import { GaxiosError } from "gaxios";
+import { google } from "googleapis";
+
+export function buildGoogleUserClient(token: GoogleUserCredential) {
+	const clientId = process.env.GOOGLE_OAUTH_CLIENT_ID;
+	if (!clientId) {
+		throw new Error("GOOGLE_OAUTH_CLIENT_ID is empty");
+	}
+	const clientSecret = process.env.GOOGLE_OAUTH_CLIENT_SECRET;
+	if (!clientSecret) {
+		throw new Error("GOOGLE_OAUTH_CLIENT_SECRET is empty");
+	}
+
+	return new GoogleUserClient(token, clientId, clientSecret);
+}
+
+// MARK: Errors
+
+class GoogleTokenRefreshError extends Error {
+	constructor(message: string, options?: { cause?: unknown }) {
+		super(message, options);
+		this.name = this.constructor.name;
+		Object.setPrototypeOf(this, GoogleTokenRefreshError.prototype);
+	}
+}
+
+/**
+ * Determines if the given error requires authorization.
+ * if return value is true, the user should be redirected to the authorization page.
+ *
+ * see {@link https://supabase.com/docs/reference/javascript/auth-signinwithoauth}
+ *
+ * @param error - The error to check.
+ * @returns True if the error requires authorization, false otherwise.
+ */
+
+export function needsAuthorization(error: unknown) {
+	if (error instanceof GoogleTokenRefreshError) {
+		return true;
+	}
+	logger.debug({ error }, "error------------");
+	if (error instanceof GaxiosError) {
+		return error.status === 401 || error.status === 403 || error.status === 404;
+	}
+	return false;
+}
+
+export type GoogleUserData = {
+	name: string;
+	email: string;
+};
+
+class GoogleUserClient {
+	private clientId: string;
+	private clientSecret: string;
+
+	constructor(
+		private token: GoogleUserCredential,
+		clientId: string,
+		clientSecret: string,
+	) {
+		this.clientId = clientId;
+		this.clientSecret = clientSecret;
+	}
+
+	async getUser(): Promise<GoogleUserData> {
+		try {
+			const authClient = await this.buildClient();
+			logger.debug(
+				{
+					hasAccessToken: !!this.token.accessToken,
+					tokenPrefix: this.token.accessToken?.substring(0, 5),
+				},
+				"getting user info",
+			);
+
+			const oauth2 = google.oauth2({
```

Comment:
[nits]I think you can build the `google.oauth2` instance inside of `buildClient()`.
However, it is trivial for now, so it's okay here.